### PR TITLE
Implement the RazorLight.Precompile tool

### PR DESCRIPTION
**Pull request by MarkKharitonov**
This method allows to implement a persistent cache as opposed to the existing in-memory cache.

I do not include an implementation of such a cache. But an example of how this can be done is:
```
public class FileSystemCachingProvider : ICachingProvider
{
    private MemoryCachingProvider m_cache = new();
    private readonly string m_root;

    public FileSystemCachingProvider(string root)
    {
        m_root = root;
    }

    public void PrecreateAssemblyCallback(IGeneratedRazorTemplate generatedRazorTemplate, byte[] rawAssembly, byte[] rawSymbolStore)
    {
        var srcFilePath = Path.Combine(m_root, generatedRazorTemplate.TemplateKey[1..]);
        var asmFilePath = srcFilePath + ".dll";
        File.WriteAllBytes(asmFilePath, rawAssembly);
        if (rawSymbolStore != null)
        {
            var pdbFilePath = srcFilePath + ".pdb";
            File.WriteAllBytes(pdbFilePath, rawSymbolStore);
        }
    }

    public void CacheTemplate(string key, Func<ITemplatePage> pageFactory, IChangeToken expirationToken)
    {
    }

    public bool Contains(string key)
    {
        var srcFilePath = Path.Combine(m_root, key);
        var asmFilePath = srcFilePath + ".dll";
        if (File.Exists(asmFilePath))
        {
            var srcLastWriteTime = new FileInfo(srcFilePath).LastWriteTimeUtc;
            var asmLastWriteTime = new FileInfo(asmFilePath).LastWriteTimeUtc;
            return srcLastWriteTime < asmLastWriteTime;
        }
        return false;
    }

    public void Remove(string key)
    {
        var srcFilePath = Path.Combine(m_root, key);
        var asmFilePath = srcFilePath + ".dll";
        var pdbFilePath = srcFilePath + ".pdb";
        if (File.Exists(asmFilePath))
        {
            File.Delete(asmFilePath);
        }
        if (File.Exists(pdbFilePath))
        {
            File.Delete(pdbFilePath);
        }
    }

    public TemplateCacheLookupResult RetrieveTemplate(string key)
    {
        var srcFilePath = Path.Combine(m_root, key);
        var asmFilePath = srcFilePath + ".dll";
        if (File.Exists(asmFilePath))
        {
            var srcLastWriteTime = new FileInfo(srcFilePath).LastWriteTimeUtc;
            var asmLastWriteTime = new FileInfo(asmFilePath).LastWriteTimeUtc;
            if (srcLastWriteTime < asmLastWriteTime)
            {
                var res = m_cache.RetrieveTemplate(key);
                if (res.Success)
                {
                    return res;
                }
                var rawAssembly = File.ReadAllBytes(asmFilePath);
                var pdbFilePath = srcFilePath + ".pdb";
                var rawSymbolStore = File.Exists(pdbFilePath) ? File.ReadAllBytes(pdbFilePath) : null;
                return new TemplateCacheLookupResult(new TemplateCacheItem(key, CreateTemplatePage));

                ITemplatePage CreateTemplatePage()
                {
                    var templatePageTypes = Assembly
                        .Load(rawAssembly, rawSymbolStore)
                        .GetTypes()
                        .Where(t => typeof(ITemplatePage).IsAssignableFrom(t) && !t.IsAbstract && !t.IsInterface)
                        .ToList();
                    if (templatePageTypes.Count != 1)
                    {
                        throw new ApplicationException($"Code bug: found {templatePageTypes.Count} concrete types implementing {nameof(ITemplatePage)} in the generated assembly.");
                    }
                    m_cache.CacheTemplate(key, CreateTemplatePage);
                    return CreateTemplatePage();

                    ITemplatePage CreateTemplatePage() => (ITemplatePage)Activator.CreateInstance(templatePageTypes[0]);
                }
            }
        }
        return new TemplateCacheLookupResult();
    }
}
```
And then it can be used like this:
```
string root = Path.GetDirectoryName(m_razorTemplateFilePath);
var provider = new FileSystemCachingProvider(root);
var engine = new RazorLightEngineBuilder()
    .UseFileSystemProject(root, "")
    .UseCachingProvider(provider)
    .AddPrecreateAssemblyCallbacks(provider.PrecreateAssemblyCallback)
    .Build();
```

### CODE REVIEW

**Comment by MarkKharitonov** [comment]
This Pull Request provides an answer to my question raised in the issue #491

File: src/RazorLight/Compilation/RoslynCompilationService.cs
Comment (jzabroski) [inline src/RazorLight/Compilation/RoslynCompilationService.cs:154]: This is bad.

File: src/RazorLight/Compilation/RoslynCompilationService.cs
Comment (jzabroski) [inline src/RazorLight/Compilation/RoslynCompilationService.cs:27]: 1. The interface here is really not "discoverable" - I have worked on this code base for over 4 years now, and just looking at this callback makes my brain hurt. The lack of a default implementation also further increases the odds people will use it incorrectly or copy-paste some bad code from somewhere on the Internet, and we will get a lot of support Issues dealing with lazy programming. - This is **the top** maintenance cost of this project thus far.
2. Why are you calling it precreate vs. precompile?
3. If your goal is to precompile assemblies, why not contribute a PR to the RazorLight.Precompile csproj?

**Review by jzabroski — CHANGES_REQUESTED** [review]
Sorry, but this isn't happening, at least not in the format you proposed. It is especially bad that you're just swallowing exceptions - that is a non-starter. The interface complexity of the callback is also very high - I don't think anyone would know what this does without reading the source code.

There is a Precompile project for compiling views, but it hasn't been finished.  Would you be interested in building a .NET Global Tool that can do that?

File: src/RazorLight/Compilation/RoslynCompilationService.cs
Comment (MarkKharitonov) [reply src/RazorLight/Compilation/RoslynCompilationService.cs:154]: In my defense I followed the exact same scheme as used for the already existing `prerenderCallbacks` - https://github.com/toddams/RazorLight/blob/22381db30ca3c61b4b1e4004e1536407dd138e49/src/RazorLight/TemplateRenderer.cs#L223-L239

I figured if that is good enough for them, why should it be different for these? However, I am will gladly change it, although I do not see any regression as opposed to the prerender callbacks.

File: src/RazorLight/Compilation/RoslynCompilationService.cs
Comment (MarkKharitonov) [reply src/RazorLight/Compilation/RoslynCompilationService.cs:27]: My reply would be this:

1. How is it different from the prerender callbacks?
2. My rationale is that it is called before the Assembly is created. The code has already been compiled, so it does not happen before the compilation. But if `precompile` is more recognizable, then it makes sense to rename. I guess this has to do with discoverability. `precompile` is a known buzz word, `precreate` is not. So, makes sense to rename. Unless, we want to keep it obscured :-).
3. I do not know. I was not aware of this tool. I did notice the project, but since it was not mentioned in the documentation (or I missed it) I just ignored it. What is the state of this tool?
4. Makes sense in general. 

TBH, I am acting on my selfish reasons - I need to implement something in Powershell yesterday and as it is either I drop RazorLight entirely (in favour of what exactly? as far as I know there is nothing being as popular and actively maintained) or complicate a straightforward and conceptually simple scripting code to move all the rendering to one executable call.

If `RazorLight.Precompile` is already usable, then definitely I would prefer using it. But it may be a longer engagement and in the meantime - the ability to cache generated assemblies is needed now. At least I need it now.

Can we find a compromise that would allow us to release something quickly yet satisfying the quality requirements? I will gladly implement all your recommendations, if at the end I will be able to achieve my goal in a timely fashion.

Hope it makes sense.

File: src/RazorLight/Compilation/RoslynCompilationService.cs
Comment (jzabroski) [reply src/RazorLight/Compilation/RoslynCompilationService.cs:154]: It's a fair point, but very few people actually utilize those callbacks because they're not easily discoverable. I've found most people follow a really old tutorial from medium.com for how to use RazorLight that is woefully out of date and contains some rather questionable decisions, that resulted in me spending way over a month figuring out why people were reporting bugs just "standing up RazorLight" that I couldn't fathom happening until I found that online tutorial's faulty guide.

File: src/RazorLight/Compilation/RoslynCompilationService.cs
Comment (jzabroski) [reply src/RazorLight/Compilation/RoslynCompilationService.cs:27]: > If `RazorLight.Precompile` is already usable, then definitely I would prefer using it. But it may be a longer engagement and in the meantime - the ability to cache generated assemblies is needed now. At least I need it now.

I'm not trying to create work for you, but it seems straight forward to lift your PR into a tool vs. something that caches everything on startup.  The RazorLight.Precompile project is just a stub right now.

> if at the end I will be able to achieve my goal in a timely fashion.

I'm not trying to create work, but I do think most people would rather precompile this stuff vs. create a persistent cache. It makes the entire flow a lot more debuggable since you have to use the compiled DLLs on disk vs. use a persistent cache that could get turned off via misconfiguration or load a different projects cached items by accident.

**Comment by MarkKharitonov** [comment]
I have changed the PR to actually implement the RazorLight.Precompile tool.
What do you think?

**Comment by MarkKharitonov** [comment]
I think it makes sense to abandon this PR and replace it with a new one.
?

**Comment by MarkKharitonov** [comment]
I have just added an ability to render the given model by the precompile command.

Please, run `precompile -h` and `render -h` to get help.

**Comment by jzabroski** [comment]
Hi Mark. I hope you are well. I have an immediate family member in the hospital so this has been on hold for now. I promise it's not simply being ignored

**Comment by MarkKharitonov** [comment]
@jzabroski - sorry to hear that. Hope things get better. Thanks for letting me know.

**Comment by jzabroski** [comment]
Getting to this now.  Expect a reply in the next 3 days.  I do think it looks good though

**Comment by jzabroski** [comment]
Thanks, Mark. Thank you as well for your patience while I worked through some things personally.

**Comment by MarkKharitonov** [comment]
No problem. Thank you for merging. Does it going to publish to nuget.org automatically?

**Comment by jzabroski** [comment]
@MarkKharitonov It looks like there is one failing test after I merged.  Can you check?  I am not immediately clear why it happened but I think its due to the backslash separator is not compatible with Ubuntu Linux?

Path separator are platform dependent :

For windows, it’s \ and for unix it’s /.


```
  Failed RenderFolderRecursive("FullMessage.cshtml","folder\\MessageItem.cshtml",RazorLight.Caching.FileHashCachingStrategy) [93 ms]
  Error Message:
   RazorLight.RazorLightException : The razor template file /home/runner/work/RazorLight/RazorLight/tests/RazorLight.Precompile.Tests/bin/Debug/net6.0/Samples/folder\MessageItem.cshtml does not exist.
  Stack Trace:
```

**Comment by MarkKharitonov** [comment]
I will have a look

**Comment by jzabroski** [comment]
https://github.com/toddams/RazorLight/blob/c7f97d6d16091046e908b7764a712d69f12767a1/tests/RazorLight.Precompile.Tests/Render2Tests.cs#L58

I think you just need a platform check.  We use xunit, which has some platform specific features built-in via plugins. But if the test is only for .NET 6, you can also just use the .NET 6 platform abstraction APIs to detect the platform elegantly.

**Comment by jzabroski** [comment]
Actually you can probably just use https://learn.microsoft.com/en-us/dotnet/api/system.io.path.pathseparator?view=netframework-4.8 since it works across all versions

**Comment by MarkKharitonov** [comment]
I need to setup a local linux environment to test, because it all passes on Windows.

**Comment by jzabroski** [comment]
I honestly dont think you need to go that far :-)  You can fork GitHub repo and set-up actions locally in your repo using our template and just make the one line change i suggested to fix the issue.  Have you ever just hit `.` on a repo after forking it? You get a VS code editor in your browser.

**Comment by MarkKharitonov** [comment]
I pushed the fix. Trying to figure out how to run the workflow. Seems like I need runners, but I do not have any yet. Will have to sort it out. 
I think there should be a PR workflow on this repository to prevent future inconveniences.

**Pull request merged**
